File: intelligence/ai_engine/discovery/data_profiler.py
"""Data profiling - samples data and computes column-level statistics."""

from typing import Any

import psycopg2
import psycopg2.extras


class DataProfiler:
    """Profiles each table: cardinality, distributions, null rates, samples."""

    def __init__(self, database_url: str):
        self.database_url = database_url

    def _connect(self):
        return psycopg2.connect(self.database_url)

    def profile_all(self, tables: list[dict], columns: dict[str, list[dict]]) -> dict[str, dict]:
        result = {}
        for table_info in tables:
            table_name = table_info["table_name"]
            if table_info.get("table_type") == "VIEW":
                continue
            table_cols = columns.get(table_name, [])
            result[table_name] = self.profile_table(table_name, table_cols)
        return result

    def profile_table(self, table_name: str, columns: list[dict]) -> dict[str, Any]:
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(f"SELECT COUNT(*) AS cnt FROM \"{table_name}\";")
                row_count = cur.fetchone()["cnt"]

                col_profiles = {}
                for col in columns:
                    col_name = col["column_name"]
                    data_type = col["data_type"]
                    col_profiles[col_name] = self._profile_column(
                        cur, table_name, col_name, data_type, row_count
                    )

                return {"row_count": row_count, "columns": col_profiles}

    def _profile_column(
        self, cur, table_name: str, col_name: str, data_type: str, total: int
    ) -> dict[str, Any]:
        safe_col = f'"{col_name}"'
        safe_table = f'"{table_name}"'

        profile: dict[str, Any] = {"data_type": data_type}

        # Null count and distinct count
        cur.execute(f"""
            SELECT
                COUNT(*) - COUNT({safe_col}) AS null_count,
                COUNT(DISTINCT {safe_col}) AS distinct_count
            FROM {safe_table};
        """)
        stats = cur.fetchone()
        profile["null_count"] = stats["null_count"]
        profile["null_rate"] = round(stats["null_count"] / total, 4) if total > 0 else 0
        profile["distinct_count"] = stats["distinct_count"]
        profile["cardinality_ratio"] = round(stats["distinct_count"] / total, 4) if total > 0 else 0
        profile["is_unique"] = stats["distinct_count"] == total and total > 0

        # Numeric stats
        if data_type in ("integer", "bigint", "smallint", "numeric", "real", "double precision"):
            try:
                cur.execute(f"""
                    SELECT
                        MIN({safe_col})::float AS min_val,
                        MAX({safe_col})::float AS max_val,
                        AVG({safe_col})::float AS mean_val,
                        STDDEV({safe_col})::float AS std_val,
                        PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY {safe_col})::float AS median_val
                    FROM {safe_table}
                    WHERE {safe_col} IS NOT NULL;
                """)
                num_stats = cur.fetchone()
                if num_stats:
                    profile["min"] = num_stats["min_val"]
                    profile["max"] = num_stats["max_val"]
                    profile["mean"] = round(num_stats["mean_val"], 4) if num_stats["mean_val"] else None
                    profile["std"] = round(num_stats["std_val"], 4) if num_stats["std_val"] else None
                    profile["median"] = num_stats["median_val"]
            except Exception:
                pass

        # Text stats
        if data_type in ("character varying", "text", "character"):
            try:
                cur.execute(f"""
                    SELECT
                        AVG(LENGTH({safe_col}))::float AS avg_length,
                        MAX(LENGTH({safe_col})) AS max_length
                    FROM {safe_table}
                    WHERE {safe_col} IS NOT NULL;
                """)
                txt = cur.fetchone()
                if txt:
                    profile["avg_text_length"] = round(txt["avg_length"], 1) if txt["avg_length"] else 0
                    profile["max_text_length"] = txt["max_length"]
                    profile["looks_like_name"] = (
                        profile["is_unique"]
                        and profile.get("avg_text_length", 0) < 80
                    )
                    profile["looks_like_description"] = profile.get("avg_text_length", 0) > 50
            except Exception:
                pass

        # Top values (for categorical detection)
        if stats["distinct_count"] <= 100 and stats["distinct_count"] > 0:
            try:
                cur.execute(f"""
                    SELECT {safe_col}::text AS val, COUNT(*) AS freq
                    FROM {safe_table}
                    WHERE {safe_col} IS NOT NULL
                    GROUP BY {safe_col}
                    ORDER BY freq DESC
                    LIMIT 10;
                """)
                profile["top_values"] = [
                    {"value": r["val"], "count": r["freq"]} for r in cur.fetchall()
                ]
            except Exception:
                profile["top_values"] = []

        # Sample values
        try:
            cur.execute(f"""
                SELECT {safe_col}::text AS val
                FROM {safe_table}
                WHERE {safe_col} IS NOT NULL
                ORDER BY RANDOM()
                LIMIT 5;
            """)
            profile["sample_values"] = [r["val"] for r in cur.fetchall()]
        except Exception:
            profile["sample_values"] = []

        return profile

File: backend/src/intelligence/discovery/dataProfiler.ts
import { QueryTypes } from 'sequelize';
import { sequelize } from '../../config/database';
import { ColumnInfo } from './schemaInspector';

export interface ColumnProfile {
  null_rate: number;
  distinct_count: number;
  cardinality: number;
  min?: number | string;
  max?: number | string;
  avg_length?: number;
}

export interface TableProfile {
  row_count: number;
  columns: Record<string, ColumnProfile>;
}

const NUMERIC_TYPES = ['integer', 'bigint', 'smallint', 'numeric', 'real', 'double precision', 'decimal'];
const TEXT_TYPES = ['character varying', 'text', 'character', 'varchar', 'char'];

function isNumeric(dataType: string): boolean {
  return NUMERIC_TYPES.some((t) => dataType.toLowerCase().includes(t));
}

function isText(dataType: string): boolean {
  return TEXT_TYPES.some((t) => dataType.toLowerCase().includes(t));
}

// Escape identifier for safe SQL interpolation
function escapeId(name: string): string {
  return `"${name.replace(/"/g, '""')}"`;
}

export async function profileTable(
  tableName: string,
  columns: ColumnInfo[],
  rowCount: number
): Promise<TableProfile> {
  if (!columns.length || rowCount === 0) {
    const emptyProfiles: Record<string, ColumnProfile> = {};
    for (const col of columns) {
      emptyProfiles[col.column_name] = {
        null_rate: 0,
        distinct_count: 0,
        cardinality: 0,
      };
    }
    return { row_count: rowCount, columns: emptyProfiles };
  }

  // Batch null/distinct counts in one query per table
  const selectParts: string[] = [];
  for (const col of columns) {
    const id = escapeId(col.column_name);
    selectParts.push(
      `COUNT(*) - COUNT(${id}) AS "null_${col.column_name}"`,
      `COUNT(DISTINCT ${id}) AS "dist_${col.column_name}"`
    );
  }

  // Add numeric min/max
  for (const col of columns) {
    if (isNumeric(col.data_type)) {
      const id = escapeId(col.column_name);
      selectParts.push(
        `MIN(${id}) AS "min_${col.column_name}"`,
        `MAX(${id}) AS "max_${col.column_name}"`
      );
    }
  }

  // Add text avg length
  for (const col of columns) {
    if (isText(col.data_type)) {
      const id = escapeId(col.column_name);
      selectParts.push(`AVG(LENGTH(${id}::text)) AS "avglen_${col.column_name}"`);
    }
  }

  const sql = `SELECT ${selectParts.join(', ')} FROM ${escapeId(tableName)}`;

  let statsRow: Record<string, any> = {};
  try {
    const [result] = await sequelize.query(sql, { type: QueryTypes.SELECT });
    statsRow = (result as Record<string, any>) || {};
  } catch {
    // Table might be empty or have permission issues
  }

  const profiles: Record<string, ColumnProfile> = {};
  for (const col of columns) {
    const nullCount = Number(statsRow[`null_${col.column_name}`]) || 0;
    const distinctCount = Number(statsRow[`dist_${col.column_name}`]) || 0;

    const profile: ColumnProfile = {
      null_rate: rowCount > 0 ? nullCount / rowCount : 0,
      distinct_count: distinctCount,
      cardinality: rowCount > 0 ? distinctCount / rowCount : 0,
    };

    if (isNumeric(col.data_type)) {
      const minVal = statsRow[`min_${col.column_name}`];
      const maxVal = statsRow[`max_${col.column_name}`];
      if (minVal != null) profile.min = Number(minVal);
      if (maxVal != null) profile.max = Number(maxVal);
    }

    if (isText(col.data_type)) {
      const avgLen = statsRow[`avglen_${col.column_name}`];
      if (avgLen != null) profile.avg_length = Math.round(Number(avgLen));
    }

    profiles[col.column_name] = profile;
  }

  return { row_count: rowCount, columns: profiles };
}

export async function profileAllTables(
  columns: Record<string, ColumnInfo[]>,
  rowCounts: Record<string, number>
): Promise<Record<string, TableProfile>> {
  const results: Record<string, TableProfile> = {};

  for (const tableName of Object.keys(columns)) {
    try {
      results[tableName] = await profileTable(
        tableName,
        columns[tableName],
        rowCounts[tableName] || 0
      );
    } catch {
      results[tableName] = { row_count: rowCounts[tableName] || 0, columns: {} };
    }
  }

  return results;
}
